### PR TITLE
ISPN-8700 ISPN-8704 ISPN-8705 Access logging fixes for Hot Rod and REST

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-8700
https://issues.jboss.org/browse/ISPN-8704
https://issues.jboss.org/browse/ISPN-8705

### CODE REVIEW

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodVersion.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+   public byte getVersion() {
+      return version;
+   }
+
+   public boolean isNewerThan(byte version) {
+      return this.version > version;
+   }
+
+   public boolean isAtLeast(byte version) {
```
Comment (wburns) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodVersion.java]: This and the `isNewerThan` methods are throwing me off. From the names I would think that both are testing this `this` object is greater than the provided version (with the latter being true when equals as well). Maybe just add some documentation to the methods to clear it up.

Maybe `isOlderOrSameThan` instead here?

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodVersion.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+   static {
+      LATEST = values()[values().length - 1];
+      Arrays.fill(versions, UNKNOWN);
+      for(HotRodVersion version : values()) {
+         versions[version.version] = version;
+      }
+   }
+
+   public static HotRodVersion forVersion(byte version) {
```
Comment (wburns) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodVersion.java]: Also it isn't clear unless I look at the source directly what format this byte should be in when it is invoked. But I guess it is fine as this is just the defined version used by the hotrod protocol.

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/logging/HotRodAccessLoggingTest.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      logAppender.install();
+      super.setup();
+   }
+
+   public void testRestAccessLog(Method m) {
+      client().put("key", "value");
+
+      String logline = logAppender.getLog(0).toString();
+      assertTrue(logline, logline.matches("^127\\.0\\.0\\.1 - \\[\\d+/\\w+/\\d+:\\d+:\\d+:\\d+ [+-]\\d+\\] \"PUT /HotRodCache/\\[B0x6B6579 HOTROD/2\\.1\" OK \\d+ \\d+ \\d+ ms$"));
+      logAppender.uninstall();
```
Comment (wburns) [inline server/hotrod/src/test/java/org/infinispan/server/hotrod/logging/HotRodAccessLoggingTest.java]: Remove this in a clean up method? If the assertion fails this logger will never get removed.

File: server/rest/src/test/java/org/infinispan/rest/logging/RestAccessLoggingTest.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      client.newRequest(String.format("http://localhost:%d/rest/default/key", restServer.getPort()))
+            .content(new StringContentProvider("value"))
+            .header("Content-type", "text/plain; charset=utf-8")
+            .method(HttpMethod.PUT)
+            .send();
+
+      String logline = logAppender.getLog(0).toString();
+
+      assertTrue(logline, logline.matches("^127\\.0\\.0\\.1 - \\[\\d+/\\w+/\\d+:\\d+:\\d+:\\d+ [+-]\\d+\\] \"PUT /rest/default/key HTTP/1\\.1\" 404 5 65 \\d+ ms$"));
+      logAppender.uninstall();
```
Comment (wburns) [inline server/rest/src/test/java/org/infinispan/rest/logging/RestAccessLoggingTest.java]: Unneeded since it is in teardown.

File: server/rest/src/test/java/org/infinispan/rest/AuthenticationTest.java [hunk 5]
```diff
@@ -52,7 +53,8 @@ public void afterMethod() throws Exception {
    public void shouldAuthenticateWhenProvidingProperCredentials() throws Exception {
       //given
       SecurityDomain securityDomainMock = mock(SecurityDomain.class, new ThrowsExceptionClass(SecurityException.class));
-      doReturn(mock(Principal.class)).when(securityDomainMock).authenticate(eq("test"), eq("test"));
+      SimpleUserPrincipal userPrincipal = new SimpleUserPrincipal("test");
```
Comment (wburns) [inline server/rest/src/test/java/org/infinispan/rest/AuthenticationTest.java]: This change seems unrelated?

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodVersion.java (hunk 1 above)
Comment (tristantarrant) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodVersion.java]: yes, I'm not a fan of the "backwards" comparison: initially I had it the other way round, converting the byte version to a HotRodVersion and comparing that to a well known version, but then that would have lost all inlining because it would become a megamorphic call... I'll come up with some better names.

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodVersion.java (hunk 2 above)
Comment (tristantarrant) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodVersion.java]: Yeah, the wire format defines the version. I'm just using simple arithmetic (major*10+minor) because that's what has been done up to now, but if we get to 2.10 we're in trouble :)

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/logging/HotRodAccessLoggingTest.java (hunk 3 above)
Comment (tristantarrant) [reply server/hotrod/src/test/java/org/infinispan/server/hotrod/logging/HotRodAccessLoggingTest.java]: Right

File: server/rest/src/test/java/org/infinispan/rest/AuthenticationTest.java (hunk 5 above)
Comment (tristantarrant) [reply server/rest/src/test/java/org/infinispan/rest/AuthenticationTest.java]: Actually no. In order to be able to log the user I had to actually return the Principa from the Authenticatorl. This broke the mock !

File: server/rest/src/test/java/org/infinispan/rest/logging/RestAccessLoggingTest.java (hunk 4 above)
Comment (tristantarrant) [reply server/rest/src/test/java/org/infinispan/rest/logging/RestAccessLoggingTest.java]: ok

**Comment by tristantarrant** [comment]
Addressed issues

**Comment by galderz** [comment]
```
[ERROR] /home/infinispan/workspace/Infinispan_PR-5687-GTBBVPVODOF5ZCORS3FFWX6UJXU3SHEOMCULP46WQMEJJ5IIWPRQ/server/rest/src/test/java/org/infinispan/rest/AuthenticationTest.java:7:8: Unused import - java.security.Principal. [UnusedImports]
```

**Comment by tristantarrant** [comment]
Fixed

**Comment by galderz** [comment]
CI failures:

```
org.infinispan.server.hotrod.logging.HotRodAccessLoggingTest.testRestAccessLog
org.infinispan.rest.logging.RestAccessLoggingTest.testRestAccessLog
```

**Comment by tristantarrant** [comment]
Fixed the regex and also stopping the server to ensure that the async logging of the netty handlers happens before the check

**Comment by gustavonalle** [comment]
looking

File: commons-test/src/main/java/org/infinispan/commons/test/skip/StringLogAppender.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * A Log4J Appender which stores logs in a StringBuilder
+ *
+ * @author Tristan Tarrant
+ * @since 9.2
+ */
+
+public class StringLogAppender extends AbstractAppender {
+   private final String category;
```
Comment (gustavonalle) [inline commons-test/src/main/java/org/infinispan/commons/test/skip/StringLogAppender.java]: Did you see ```org.apache.log4j.WriterAppender``` ?

File: server/core/src/main/java/org/infinispan/server/core/logging/LogUtil.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      moy.put(9L, "Sep");
+      moy.put(10L, "Oct");
+      moy.put(11L, "Nov");
+      moy.put(12L, "Dec");
+      LOG_DATE_TIME_FORMATTER = new DateTimeFormatterBuilder()
+            .parseCaseInsensitive()
+            .parseLenient()
+            .appendValue(DAY_OF_MONTH, 1, 2, SignStyle.NOT_NEGATIVE)
+            .appendLiteral('/')
+            .appendText(MONTH_OF_YEAR, moy)
```
Comment (gustavonalle) [inline server/core/src/main/java/org/infinispan/server/core/logging/LogUtil.java]: Questions: 
1) in the server config file, there are handlers ```HR-ACCESS-FILE``` and ```REST-ACCESS-FILE``` that apparently allows you to define the ```pattern-formatter``` for the access log, but those are actually ignored in favor of this harcoded format?
2) What about i18 here? Are going to a fixed date format with month names in english?

**Comment by gustavonalle** [comment]
@tristantarrant probably not related to this PR, but since you're at it: I noticed the access log is written to *-acess.log but also to server.log. I think it shouldn't, thoughts?

**Comment by gustavonalle** [comment]
LGTM, have a couple of questions though

File: commons-test/src/main/java/org/infinispan/commons/test/skip/StringLogAppender.java (hunk 6 above)
Comment (tristantarrant) [reply commons-test/src/main/java/org/infinispan/commons/test/skip/StringLogAppender.java]: Damn. I'll use that.

**Comment by tristantarrant** [comment]
Ok, rethinking this: I will use the MDC and provide a logging pattern that mimics the CLF or NCSA log format.

File: commons-test/src/main/java/org/infinispan/commons/test/skip/StringLogAppender.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.logging.log4j.core.config.LoggerConfig;
+
+/**
+ * A Log4J Appender which stores logs in a StringBuilder
+ *
+ * @author Tristan Tarrant
+ * @since 9.2
+ */
+
+public class StringLogAppender extends AbstractAppender {
```
Comment (gustavonalle) [inline commons-test/src/main/java/org/infinispan/commons/test/skip/StringLogAppender.java]: Forgot to get rid of this?

Comment (tristantarrant) [reply commons-test/src/main/java/org/infinispan/commons/test/skip/StringLogAppender.java]: Didn't find a way to implement a thread id filter.

Comment (gustavonalle) [reply commons-test/src/main/java/org/infinispan/commons/test/skip/StringLogAppender.java]: ok, no problems!

**Comment by gustavonalle** [comment]
Looks nice 👍 
waiting for CI to merge it

**Comment by gustavonalle** [comment]
merged, tks @tristantarrant
